### PR TITLE
FTX - implement trigger-order parsing

### PR DESCRIPTION
**Pull request by ttodua**
Implements fetching & parsing the conditional orders (_this PR also fixes #12523_). Before this, it was not possible to `fetchOrder` conditional order, as FTX doesn't have native endpoint to fetch single trigger order (using `/orders/{order_id}` , as it seems an exception from their side, as there are all other needed endpoints for trigger orders).
At first glance, this implementation seems imperfect, and I agree, but on the second hand, it covers probably 99.9$ use-cases. This PR implements two possible ways to fetch conditional order:
1) first way is  :
```
const myOrder = await e.fetchOrder(id, symbol, {type:'takeProfit'});
```
In the backgrounds fetches all open trigger orders **by symbol** (symbol is requirement) using endpoint `/conditional_orders?market={market}`.  ftx provides at least 100 orders when fetching from open orders. so, on same symbol, it is nearly impossible that anyone on earth had more than 100 open **trigger orders** on same symbol. it is just impossible and such erroneous cases will never be faced. thus, after fetching all open **trigger orders**, among them we can easily return the order with the requested `id` passed through `fetchOrder (id, ...). I know, this is not ideal, but does the job and probably covers many users needs. and even more - it's better than no ability (for conditional-orders)  to have unified ccxt way to fetchOrder  at all .

2) second way is:
```
const myOrder = await e.fetchOrder(id, symbol, {type:'takeProfit', method:'privateGetConditionalOrdersConditionalOrderIdTriggers'});
```
which returns the object if the trigger-order has been executed (even partially) and it's easy to detect & parse it. Either this way, user can get the basic properties from 'order' object (id, reference-order id, order-status, order-amount & filled amount). Still, it's better than no ability (for conditional-orders) to have unified ccxt way to fetchOrder  at all .

### CODE REVIEW

**Comment by ttodua** [comment]
Seems slight issue in my PR. will commit another one

**Comment by ttodua** [comment]
I've removed the `1)` approach, and now only `2)` approach can be used:
```
const myOrder = await e.fetchOrder('173009180', 'BTC-PERP', {type:'takeProfit', method:'privateGetConditionalOrdersConditionalOrderIdTriggers'});
```
and returns minimal (whatever available) properties for that order.

**Comment by ttodua** [comment]
Now it's also possible to make `fetchOrderTrades (id, ...., {type:'takeProfit'}`

**Comment by ttodua** [comment]
@kroitor 
I think I've implemented all aspects I could do with current update. 
there still could be made some improvement in future (if ftx will fix api and allow to query for conditional-order normally, with order-id).

also, second thing : I can implement `fetchOpenOrder(id...` which will require 'symbol' argument to be provided by user. in such case, we can iterate through `fetchOpenOrders(symbol...)` returned items (which is at least 100 entries) and from there, match the order-id and return that object from singular `fetchOpenOrder(id)` call. it is for sure that noone would have more than 100 open orders for specific symbol, so this will never have any issue/pagination glitch etc...
what do you think?

**Comment by ttodua** [comment]
p.s just for info: so, in FTX, there are 2 kinds of orders (returned from API responses):
`(A)` regular order (with 'regular' structure, which ), which has ID length 12 digits _(by this moment)_,  You can retrieve them by  regular `fetchOrders (....,)`, which goes to `/orders/history` endpoint.
`(B)` conditional (a.k.a. trigger) order (with 'extended' structure), which has ID length 9 digits .  You can retrieve them by i.e.  `fetchOrders (...., {type:'takeProfit'} )`, which goes to `/condiitonal_orders/history` endpoint.
[see [diff](https://www.diffchecker.com/PaSkE1Ni) ]

However, when you obtain 'orders' from fetchOrders, then with regular-order's id (12 digits) you can retrieve-back (with `fetchOrder(id)` which goes to `/orders/{order_id}` endpoint) the same singular order-object, but with conditional-order's `id` (which has 9 digits) you can't retrieve that singular conditional-order object back :) instead, with that `id` you can retrieve an object (from `/triggers` endoint), which has referenced `orderId` (12 digits) which can be used to retrieve that singular order object using `fetchOrder(orderId)` (from `/orders/{order_id}` endpoint) - however, that will have the same structure as 'regular' order, and not the `(B)` structure.
very complicated story :)

**Comment by ttodua** [comment]
back to PR: I think I've implemented current solution for fetching order with singular 'conditional-order' id.
there still could be made some improvement in future (if ftx will fix api and allow to query for conditional-order normally, with order-id) or I can implement fetchOpenOrder(id... which will require 'symbol' argument to be provided by user. in such case, we can iterate through fetchOpenOrders(symbol...) returned items (which is at least 100 entries) and from there, match the order-id and return that object from singular fetchOpenOrder(id) call. it is for sure that noone would have more than 100 open orders for specific symbol, so this will never have any issue/pagination glitch etc...
what do you think?

**Comment by micant** [comment]
Thanks for putting together this PR so quickly and for your discovery work on the problem - massive help.

> it is for sure that noone would have more than 100 open orders for specific symbol, so this will never have any issue/pagination glitch etc...
> what do you think?

I agree. I can't imagine someone having 100+ open trigger orders and if they do, at least the alternative path is documented 👍  
This looks like a solid solution until FTX bridges the gap.

**Comment by ttodua** [comment]
@kroitor I've tested manually to create &fetch orders, and seems this is working and doesn't have any possible side-effects. (mostly, as it's in other implementations of ftx methods, i add checking for `type` param, and when it's a trigger type, act as appopriate. otherwise, fallback to previous functionalilty). However, lmk your thoughts.

File: js/ftx.js [hunk 1]
```diff
@@ -1695,42 +1714,91 @@ module.exports = class ftx extends Exchange {
...
-        const response = await this[method] (this.extend (request, params));
+        const type = this.safeValue (params, 'type');
+        params = this.omit (params, 'type');
+        // Note, the below types use non-standard endpoints, look through the implementation to understand better (It might better to use `fetchOrders/fetchOpenOrders` to get your trigger order's details)
+        if ((type === 'stop') || (type === 'trailingStop') || (type === 'takeProfit')) {
+            const orderStatusFlag = this.safeString (params, 'orderStatusFlag');
+            params = this.omit (params, 'orderStatusFlag');
+            if (method === 'privateGetConditionalOrdersConditionalOrderIdTriggers') {
+                const order = await this.fetchConditionalTriggersHelper (id, params);
+                order['trigger_order_flag'] = true;
```
Comment (kroitor) [inline js/ftx.js:1738]: This is a no-go, we should not inject this field into the raw order to be parsed. The order must remain untouched, so no injections and no removals. Also, if possible, please, remove the fetchConditionalTriggerHelper, it does not really save us much space, but makes the code harder to follow. Ideally, please solve this task without introducing new methods or fields into the equation.

File: js/ftx.js [hunk 2]
```diff
@@ -1378,11 +1378,30 @@ module.exports = class ftx extends Exchange {
         //         "type":"limit"
         //     }
         //
+        // fetchOrder (condition-order)
+        //
+        //     {
+        //         "time": "2022-03-29T04:54:04.390665+00:00",
+        //         "orderId": 132144259673,  //this is not the same id used in fetchOrder, instead reference id
```
Comment (kroitor) [inline js/ftx.js:1385]: A side-note on formatting, please, always put one space (not two) before the opening `//` comment, and one space after. Keep the formatting uniform across the file and across the exchanges.

```JavaScript
"orderId": 12345678,  //this is not the same id used in fetchOrder, instead reference id
↓
"orderId": 12345678, // this is not the order id used in fetchOrder, it is a reference id
```

File: js/ftx.js (hunk 1 above)
Comment (ttodua) [reply js/ftx.js]: considered all notes, just I've one question:
`fetchConditionalTriggersHelper` is not being used alone on that line, but that is being used from 2 different places (that's why i thought to put in a helper method, as the code was reused. also, i've included special comments in that helper method, to explain the related endpoint's behavior there). should I still remove that helper and place two copies of that code-block (with same comments) into those two places?

**Pull request closed without merging**